share/p2p/middleware.go

```diff
@@ -9,21 +9,35 @@ import (
 
 var log = logging.Logger("shrex/middleware")
 
-func RateLimitMiddleware(inner network.StreamHandler, concurrencyLimit int) network.StreamHandler {
-	var parallelRequests int64
-	limit := int64(concurrencyLimit)
+type Middleware struct {
+	// NumRateLimited is the number of requests that were rate limited. It is reset to 0 every time
+	// it is read and observed into metrics.
+	NumRateLimited atomic.Int64
```

Comment:
There is no need to export it if you embed metrics struct with observe methods same way you done with shrex client / server. Or maybe there is a reason why you would want to move observation from this module to server?

Reply:
Its a bit more tricky than that unfortunately. I can either export a method on the middleware, but that method ofc doesnt have access to both eds server and nd server's metrics. So I decided to expose the counter instead. Would you rather it be an exposed method that drains the counter and returns the result? This is the code im talking about (exists in both servers):

```golang
func (srv *Server) observeRateLimitedRequests(ctx context.Context) {
	var numRateLimited int64
	if srv.metrics != nil {
		numRateLimited = srv.middleware.NumRateLimited.Swap(0)
	}

	if numRateLimited > 0 {
		srv.metrics.ObserveRequests(ctx, numRateLimited, p2p.StatusRateLimited)
	}
}
```

Reply:
After discussion w @walldiss, added #2157 and settled on exporting a method for now to get this into release